src/main/java/io/jenkins/plugins/credentials/secretsmanager/AwsCredentialsSnapshotTaker.java

```diff
@@ -0,0 +1,20 @@
+package io.jenkins.plugins.credentials.secretsmanager;
+
+import com.cloudbees.plugins.credentials.CredentialsSnapshotTaker;
+
+import hudson.Extension;
+
+@Extension
+@SuppressWarnings("unused")
+public class AwsCredentialsSnapshotTaker extends CredentialsSnapshotTaker<AwsCredentials> {
```

Comment:
https://github.com/jenkinsci/azure-keyvault-plugin/pull/96#issuecomment-998119909

I see that your ITs directly call `CredentialsProvider.snapshot`, but that is not really testing whether an agent build would do the same.

Reply:
I believe that calling snapshot is something the downstream credentials consumer plugin has to do. As previously mentioned this isn't ideal; the main consumer plugins like pipeline do this, but other plugins don't.

Would there be a central higher-level place, maybe in the Credentials API, where we could add something to always snapshot the retrieved credential if a plugin is running in a remote agent context?

Reply:
> the main consumer plugins like pipeline do this, but other plugins don't

AFAICT only `git-client` does it, because the weird implementation of `git-client` + `git` sends a remoted proxy object for all operations. (Other SCM plugins get by without such tricks.) I have not come across other plugins which serialize `Credentials` instances over a Remoting channel but it would be hard to search for this pattern.

> we could add something to always snapshot the retrieved credential if a plugin is running in a remote agent context

I suppose `BaseCredentials` could implement `writeReplace` for this purpose (taking care to differentiate Remoting from XStream serialization). I am not sure what such a change could break. I came across a `CredentialsProvider.snapshot` call site in CloudBees CI written by the author of `CredentialsSnapshotTaker` which makes me nervous about the compatibility implications, unless `CredentialsSnapshotTaker` implementations are always idempotent.

Ideally integration tests would exercise authenticated Git checkouts. That could be pretty tricky though. Even assuming e.g. a WireMock for Secrets Manager, `GitSampleRepoRule` does not support authentication.